algorithms/linfa-clustering/src/k_means/hyperparameters.rs

```diff
@@ -120,23 +62,73 @@ impl<F: Float, R: Rng + Clone> KMeansHyperParams<F, R> {
     ///   in terms of inertia that the ones which minimizes the sum of squared
     ///   euclidean distances to the closest centroid for all observations.
     ///
-    /// `n_clusters` is mandatory.
-    ///
     /// Defaults are provided if optional parameters are not specified:
-    /// * `tolerance = 1e-4`;
-    /// * `max_n_iterations = 300`.
-
-    pub fn new_with_rng(n_clusters: usize, rng: R) -> KMeansHyperParamsBuilder<F, R> {
-        KMeansHyperParamsBuilder {
+    /// * `tolerance = 1e-4`
+    /// * `max_n_iterations = 300`
+    /// * `n_runs = 10`
+    /// * `init = KMeansPlusPlus`
+    pub fn new(n_clusters: usize, rng: R, dist_fn: D) -> Self {
+        Self {
             n_runs: 10,
             tolerance: F::cast(1e-4),
             max_n_iterations: 300,
             n_clusters,
             init: KMeansInit::KMeansPlusPlus,
             rng,
+            dist_fn,
+        }
+    }
+
+    /// Change the value of `n_runs`
+    pub fn n_runs(mut self, n_runs: usize) -> Self {
+        self.n_runs = n_runs;
+        self
+    }
+
+    /// Change the value of `tolerance`
+    pub fn tolerance(mut self, tolerance: F) -> Self {
+        self.tolerance = tolerance;
+        self
+    }
+
+    /// Change the value of `max_n_iterations`
+    pub fn max_n_iterations(mut self, max_n_iterations: u64) -> Self {
+        self.max_n_iterations = max_n_iterations;
+        self
+    }
+
+    /// Change the value of `init`
+    pub fn init_method(mut self, init: KMeansInit<F>) -> Self {
+        self.init = init;
+        self
+    }
+
+    /// Return an instance of `KMeansHyperParams` after
+    /// having performed validation checks on all the specified hyperparameters.
+    pub fn build(self) -> Result<KMeansHyperParams<F, R, D>, KMeansParamsError> {
```

Comment:
any reason why not call this `validate` and perform the check once fitting starts?

Reply:
If I do that then I have to call `validate` in both `fit` and `fit_with` instead of just one place. Also calling `validate` in every iteration of `fit_with` is inefficient.

Reply:
no for me lazy validation seems the right approach, because
 * the validation cost is negligible compared to the training (and `fit_with` can be called with update hyperparameters)
 * the use of `build` prior to training/prediction looks cluttered to me